[TypeChecker] Prevent pattern resolution from triggering type-checking for unresolved references

xedin opened the pull request:
Workaround for Clang enum typaliases called `isInvalid()` and `getInterfaceType()`
on discovered declarations. That doesn't play well with result builders because
declarations/patterns used in the body of a result builder don't get a type until
a solution is applied. Calling `isInvalid()` or `getInterfaceType` on declarations
located in the body of a result builder trigger a separate type-check for the
declaration that interferes with builder transformation and leads to crashes.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[BuilderTransform] NFC: Print out a function to which builder is being applied

xedin committed:
[TypeChecker] Prevent pattern resolution from triggering type-checking for unresolved references

Workaround for Clang enum typaliases called `isInvalid()` and `getInterfaceType()`
on discovered declarations. That doesn't play well with result builders because
declarations/patterns used in the body of a result builder don't get a type until
a solution is applied. Calling `isInvalid()` or `getInterfaceType` on declarations
located in the body of a result builder trigger a separate type-check for the
declaration that interferes with builder transformation and leads to crashes.

xedin commented:
@swift-ci please test source compatibility

hborla commented on lib/Sema/BuilderTransform.cpp:
```diff
@@ -1661,6 +1661,14 @@ Optional<BraceStmt *> TypeChecker::applyResultBuilderBodyTransform(
+  if (cs.isDebugMode()) {
+    auto &log = llvm::errs();
+
+    log << "--- Applying result builder to function ---\n";
```
Nice, this will be super helpful in debug-constraints output 🙂

xedin replied:
Yeah, it's going to be a good indicator what is being processed piecemeal, otherwise it's sometimes unclear where the sub-targets are coming from...

xedin commented:
@swift-ci please smoke test Windows platform

xedin commented:
I have changed to check for `hasInterfaceType()` and left a comment that explains why this is necessary what what happens if we simply remove the check.

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test Windows platform

xedin committed:
[TypeChecker] Remove `isInvalid` and related assertion from `filterForEnumElement`

Since re-declaration error is going to be emitted anyway it doesn't
matter which declaration would get returned.

xedin commented:
@swift-ci please smoke test

xedin commented:
As suggested by @slavapestov removed both `isInvalid` check and related `assertion`.
